### DISCUSSION TITLE
Trusted lists in uBO

### DISCUSSION THREAD

**Discussion by ryanbr** [uBlockOrigin/uAssets#18589] [General]
Testing trusted-set-cookie, and then finding out uBO can't use any filters in Easylist Cookie/Annoyances, Can we cover more these lists as "trusted".

https://github.com/easylist/easylist/commit/10770c57d9f3e90e1c23fe7163af8f4441e2eb49

**Comment by ryanbr** [comment]
It would expand and fix many exception rules used in ELC and Annoyances (that set cookies). Nothing nefarious.

**Comment by uBlock-user** [comment]
Currently, there are only two sources of TRUSTED filters - 

- my Filters

- lists hosted at https://github.com/uBlockOrigin/uAssets

In the future, gorhill has said he may bring support for third-party lists being **manually** set to TRUSTED by uBO users.

**Reply by ryanbr** [reply]
These aren't new subscriptions, or hosted on some random servers. Asking to re-evaluate the decision to 'paywall it off. @gorhill

**Reply by uBlock-user** [reply]
If you're asking for a TRUSTED by default, I doubt that will happen, the whole point of TRUSTED filter lists is to limit and control who can make use of such scriptlets.

**Reply by ryanbr** [reply]
I'm talking about trusted sources included in uBlock Origin. Just ELC/FB Annoyances could take adavantage of these. Improving reliablility of blocking, that would be a win for users.

With any lists included with uBO there must be some level of "trust", without it uBO blocking would be limited to its own lists.

I'm not trying to argue here, I'm just trying to change the decsion on what is "trusted".

**Reply by MasterKia** [reply]
> some level of "trust"

How can a list be trusted by uBO when the uBO team has no write access to that list?

**Reply by ryanbr** [reply]
uBO includes many different (annoyances, regionals..) lists, and only handled by specific list author or authors. Including the list as part of uBO indicates some level of trust.  Which is what I'm saying.

**Reply by Yuki2718** [reply]
Of note, we discussed very this **before** the first trusted scriptlet was added. This discussion rather sounds minus to us, because it's as if you underestimate the risk of those scriptlet.

**Reply by Yuki2718** [reply]
> I'm just trying to change the decsion on what is "trusted"

The term "trusted" is misleading, we just followed AG's term but it does not mean we trust ourselves, no, we should not. There's no absolute way to prevent misuse of trusted scriptlets, e.g. trusted-click-element can be misused once a rules is made by anyone who has only partial access to the site, which has repeatedly been discussed since jspenguin added the clicker. Untrusted sciptlets can only break sites or just weaken site's security at worst, not at all comparable to the possible consequences of trusted scriptlets. What we can do is to minimize use itself and also access to the scriptlet; i.e. risk control. Having mutual review is expceted to reduce chance of unnecessary use, e.g. cookie set by a site having random value does not mean we need trusted-set-cookie as sometimes `1` works regardless of actual value. This is a matter of [security model](https://github.com/uBlockOrigin/uBlock-issues/issues/1845#issuecomment-979275035) and possibly the one of responsibility.

**Comment by ryanbr** [comment]
I'm only talking in the best interests of users here, we have many [exception rules](https://github.com/easylist/easylist/blob/master/easylist_cookie/easylist_cookie_allowlist_general_hide.txt) that I would like to fix up.  I already started to implement fixes https://github.com/easylist/easylist/blob/master/easylist_cookie/easylist_cookie_specific_uBO.txt#L285

I don't want Easylist Cookie to be a 2nd class citizen on uBO because of politics and hypothticals. Asking as a list author, Just allow certain urls/mirrors access to trusted elements to improve ELC/Annoyances further.

@gorhill

**Reply by stephenhawk8054** [reply]
Outside of (what you assumed) "politics and hypotheticals", within the view of list authors, if years-long contributors (mapx, okiehsch...) and/or other active contributors (Yuki2718...) want to have write access to EL repo, would EL agree or still restrict to current few people?

**Reply by ryanbr** [reply]
How is that a requirement? or am I missing something

**Reply by okiehsch** [reply]
> I don't want Easylist Cookie to be a 2nd class citizen on uBO because of politics and hypothticals. 

What politics? I get that security concerns are always "hypothetical", until their not, but why do you think the limited access to those scriptlets was implemented other than those security concerns.

**Reply by stephenhawk8054** [reply]
> How is that a requirement? or am I missing something

It's as what @MasterKia has said. These trusted scriptlets are dangerous, and I'm not being hypotheticals, politics or anything. They literally can create any arbitrary new contents of its kind (whether it's about document source, scripts, cookies, local storage...). These scriptlets are what uBO team agreed on when creating/adapting them, therefore the concept of trusting here is completely different to other kinds of filters before. It's not about what lists are currently be default here. If the scriptlets are on a list, at least some team members should have control within it. These are the ones we have discussed and debated for very long time, about its dangerousness and the security-risk on uBO, or how we should craft the filters.

If the above is not possible then we are totally fine. And just for your information, the `https://github.com/uBlockOrigin/uAssets/blob/master/filters/annoyances-cookies.txt` list could be merged to ELC version in uBO by default (which means when users enable ELC, those fixes with trusted scriptlets are automatically in there). That is the reason why it was created. It's similar to how other lists in uBO extend the fix with uBO's syntaxes. You can create PR for those cookie issues if you want to fix any sites.

**Reply by Yuki2718** [reply]
> I already started to implement fixes 

An accomplished fact does not help here. On Slack and here you keep showing a complete indifference about security risk they have and looking only potential utility, which is a red flag and we admonished ourselves for. Remember, (some of) trusted scriptlets can be abused even if the rule is made by the good hand - we can only control the degree of damage and no more potential incident itself. I said almost the same in our internal discussion too.

**Reply by Yuki2718** [reply]
> the best interests of users

How much? In AG Cookie where trusted-set-cookie has been used for months, I only see 20+ rules of it. Generally, websites really needing trusted scriptlets are rare.

**Reply by gorhill** [reply]
@ryanbr The idea of having `trusted-` scriplets being valid only in uBO's own filter lists is that any additions/modifications of filters are made under the eyes of all volunteers here, something which is not possible with lists outside this repo.

Other blockers also have such requirements when it comes to security- or privacy-sensitive fitlers, this is not specific to uBO. This won't change.

What can be done though is submit pull requests to the repo here the same way volunteers here submit pull requests to EasyList repo when they want to submit a fix to EasyList-related lists.

---

Also, if not a PR, just a cc to someone here with a suggested fix is also fine to reduce overhead.

**Reply by gorhill** [reply]
@ryanbr This is how it should look in dev build once it updates with latest `assets.dev.json`:

![b](https://github.com/uBlockOrigin/uAssets/assets/585534/656192d3-4083-413a-b725-ee2ece46ca67)

**Comment by dportvine** [comment]
@ryanbr 
> I'm only talking in the best interests of users here

I just want you to know that. EasyList – Cookies Notices removed from uBOLite. 
https://github.com/gorhill/uBlock/commit/4ddfb3f2789d8fe48b62886af74184262c05d29b
https://github.com/easylist/easylist/commit/583f37786fccc19e7afbbac4414845fd511f4dd3
https://github.com/AdguardTeam/AdguardFilters/commit/fd9f413ef4e3e7890526131023441ed46e79d89e

**Reply by ryanbr** [reply]
If @gorhill wants to limit our lists, not much I can do. I have no time for uBO politics anymore

**Reply by gorhill** [reply]
When you bring terms like "politics" (whatever that means, nobody understand here) in the discussion, you poison the well, you make it impossible to cleanly discuss technical issues as you preemptively reject that in the end it's always possible to disagree without ill intents.

I moved to AdGuard Cookies because of the [issue](https://github.com/uBlockOrigin/uBOL-issues/issues/5#issuecomment-1583192594) that was lingering unresolved. Lists do not update in uBOL, so I had to go with the lists for which I knew the issue was resolved, and the change I made takes into account that I may switch back to EasyList, as I neutralized the list id I assigned to lists ( i.e. `annoyances-cookies` instead of `adguard-ccokies` or `easylist-cookies`).

It looks like you resolved it a few hours before my change, but I was never notified of this. and I can't start spending my time scanning all commits to EasyList to find out whether something I flagged through "cc @ryanbr" has been fixed.

**Reply by ryanbr** [reply]
I'm sorry it took longer than expected, but I get so many requests each day, every day. Miss just one request in a day, the following requests come through and so fourth. Prodding via Slack is often quicker, if its something urgent. I respond and or at least fix every uBO author issue or request.  Where do I poison anything?

**Reply by ryanbr** [reply]
I literally have no input before things occur in uBO related to the lists we manage, they just happen after the fact when I'm notified. I miss just one notification in error and there are sudden consequences.

**Reply by ryanbr** [reply]
What can we or I do to resolve the issue, put everyone on friendlier terms.

**Reply by gorhill** [reply]
For any `trusted-...` scriptlets, they need to be committed in [uAssets](https://github.com/uBlockOrigin/uAssets). For example, for cookie-related stuff, this has to go in `https://github.com/uBlockOrigin/uAssets/blob/master/filters/annoyances-cookies.txt`, so as mentioned above, either a (bulk) PR or "cc @...". It's about having very strict rules regarding all `trusted-` scriptlets, if we start to make exceptions than it becomes arbitrary from an outsider's point of view.

Beside sticking to strict rules, other reasons to not allow `trusted-` scriptlets in other repos is to have full control of usage of these, and it's in line with other blockers. Should an issue arise in the wild with a trusted scriptlet, we can act immediately because they are all present in uAssets.

As shown in the screenshot above, the `trusted-...` scriptlets are virtually part of EasyList from a user's point of view, so _EasyList -- Cookie Notices_ is not relegated to _2nd class citizen_, someone selecting _EasyList/uBO -- Cookie Notices_ will benefit from the `trusted-` scriptlets.

**Reply by ryanbr** [reply]
RIght. Can we fix up MV3 Easylist Cookie then?

**Reply by gorhill** [reply]
<https://github.com/gorhill/uBlock/commit/6a817b82587cff0d942b158a2cf053e51015475b>, change effective in next version of uBOL.

**Reply by Yuki2718** [reply]
> I'm sorry it took longer than expected, but I get so many requests each day, every day. 

Actually we discussed this too, this was another context of https://github.com/uBlockOrigin/uAssets/discussions/18589#discussioncomment-6226892

**Reply by ryanbr** [reply]
Its a bit late here, I'll organise with the authors the best way to up the commits.
